### PR TITLE
MVPoly: implement monomials form

### PR DESCRIPTION
**Pull request by dannywillems**
It aims to share the same interface as `prime::Dense`, but it is a WIP. I would suggest to add a trait later.

Built on top of #2514

### CODE REVIEW

File: mvpoly/src/monomials.rs [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Add<&Sparse<F, N, D>> for Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn add(self, other: &Sparse<F, N, D>) -> Self::Output {
```
Comment (marcbeunardeau88) [inline mvpoly/src/monomials.rs:54]: isn't the same function implemented l28 ?

File: mvpoly/src/monomials.rs [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Add<Sparse<F, N, D>> for &Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn add(self, other: Sparse<F, N, D>) -> Self::Output {
```
Comment (marcbeunardeau88) [inline mvpoly/src/monomials.rs:80]: same comment

File: mvpoly/src/monomials.rs [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Add<&Sparse<F, N, D>> for &Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn add(self, other: &Sparse<F, N, D>) -> Self::Output {
```
Comment (marcbeunardeau88) [inline mvpoly/src/monomials.rs:106]: same comment

File: mvpoly/src/monomials.rs [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+/// Equality is defined as equality of the monomials.
+impl<const N: usize, const D: usize, F: PrimeField> PartialEq for Sparse<F, N, D> {
+    fn eq(&self, other: &Self) -> bool {
+        self.monomials == other.monomials
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Eq for Sparse<F, N, D> {}
```
Comment (marcbeunardeau88) [inline mvpoly/src/monomials.rs:322]:
```suggestion

```

File: mvpoly/src/monomials.rs [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Zero for Sparse<F, N, D> {
+    fn is_zero(&self) -> bool {
+        self.monomials.len() == 1
+            && self.monomials.contains_key(&[0; N])
+            && self.monomials[&[0; N]].is_zero()
+    }
+
+    fn zero() -> Self {
```
Comment (marcbeunardeau88) [inline mvpoly/src/monomials.rs:339]: why is zero not the empty map ?

**Review by marcbeunardeau88 — APPROVED** [review]

File: mvpoly/src/monomials.rs (hunk 1 above)
Comment (dannywillems) [reply mvpoly/src/monomials.rs:54]: :sweat_smile: 
We can use macros, but it is not really readable.
There might also be a better way that I don't know.

File: mvpoly/src/monomials.rs (hunk 4 above)
Comment (dannywillems) [reply mvpoly/src/monomials.rs:322]: No, it is necessary to implement the trait `Eq`.

File: mvpoly/src/monomials.rs (hunk 5 above)
Comment (dannywillems) [reply mvpoly/src/monomials.rs:339]: I wanted to have a single representation for the zero polynomial, and I picked the constant polynomial with `0`. No strong opinion. Feel free to change it later.

**Pull request merged**